clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/LoadBalancerUpsertHandler.groovy

```diff
@@ -0,0 +1,115 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.aws.deploy.handlers
+
+import com.amazonaws.AmazonServiceException
+import com.amazonaws.services.elasticloadbalancing.AmazonElasticLoadBalancing
+import com.amazonaws.services.elasticloadbalancing.model.ApplySecurityGroupsToLoadBalancerRequest
+import com.amazonaws.services.elasticloadbalancing.model.CreateLoadBalancerListenersRequest
+import com.amazonaws.services.elasticloadbalancing.model.CreateLoadBalancerRequest
+import com.amazonaws.services.elasticloadbalancing.model.DeleteLoadBalancerListenersRequest
+import com.amazonaws.services.elasticloadbalancing.model.Listener
+import com.amazonaws.services.elasticloadbalancing.model.LoadBalancerDescription
+import com.netflix.spinnaker.clouddriver.data.task.Task
+import com.netflix.spinnaker.clouddriver.data.task.TaskRepository
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperationException
+
+class LoadBalancerUpsertHandler {
+
+  private static final String BASE_PHASE = "UPSERT_ELB"
+
+  private static Task getTask() {
+    TaskRepository.threadLocalTask.get()
+  }
+
+  public static void updateLoadBalancer(AmazonElasticLoadBalancing loadBalancing, LoadBalancerDescription loadBalancer,
+                                  List<Listener> listeners, Collection<String> securityGroups) {
+    def amazonErrors = []
+    def loadBalancerName = loadBalancer.loadBalancerName
+
+    if (listeners) {
+      def existingListeners = loadBalancer.listenerDescriptions*.listener.findAll {
+        // ignore the old listener :0 => :0, which AWS adds to ELBs created sometime before 2012-09-26
+        it.protocol && it.instancePort != 0 && it.loadBalancerPort != 0
+      }
+      def listenersToRemove = existingListeners.findAll {
+        // existed previously but were not supplied in upsert and should be deleted
+        !listeners.contains(it)
+      }
+      listeners.removeAll(listenersToRemove)
+
+      // no need to recreate existing listeners
+      listeners.removeAll(existingListeners)
+
+      listenersToRemove.each {
+        loadBalancing.deleteLoadBalancerListeners(
+          new DeleteLoadBalancerListenersRequest(loadBalancerName, [it.loadBalancerPort])
+        )
+        task.updateStatus BASE_PHASE, "Listener removed from ${loadBalancerName} (${it.loadBalancerPort}:${it.protocol}:${it.instancePort})."
+      }
+
+      // ignore the old listener :0 => :0, which AWS adds to ELBs created sometime before 2012-09-26
+      listeners
+        .findAll { it.instancePort != 0 && it.loadBalancerPort != 0 && it.protocol }
```

Comment:
ah wait I see its not, its just the same check - maybe just move the filtering out of that one above the existingListener check so its only in one spot though?